pkg/cluster/resources.go

```diff
@@ -135,18 +135,14 @@ func CreateNamespace(cli client.Client, namespace string, metaOptions ...MetaOpt
 	}
 
 	foundNamespace := &corev1.Namespace{}
-	err := cli.Get(context.TODO(), client.ObjectKey{Name: namespace}, foundNamespace)
-	if err != nil {
-		if apierrs.IsNotFound(err) {
-			err = cli.Create(context.TODO(), desiredNamespace)
-			if err != nil && !apierrs.IsAlreadyExists(err) {
-				return nil, err
-			}
-			desiredNamespace.DeepCopyInto(foundNamespace)
-		} else {
-			return nil, err
-		}
+	if getErr := cli.Get(context.TODO(), client.ObjectKey{Name: namespace}, foundNamespace); client.IgnoreNotFound(getErr) != nil {
+		return nil, getErr
+	}
+
+	createErr := cli.Create(context.TODO(), desiredNamespace)
+	if apierrs.IsAlreadyExists(createErr) {
+		return foundNamespace, nil
 	}
 
-	return foundNamespace, nil
+	return desiredNamespace, client.IgnoreAlreadyExists(createErr)
```

Comment:
I would say for me straight forward logic is more clear, like:
```
func CreateNamespace(cli client.Client, namespace string, metaOptions ...MetaOptions) (*corev1.Namespace, error) {
	ns := &corev1.Namespace{}
	err := cli.Get(context.TODO(), client.ObjectKey{Name: namespace}, ns)
	if err == nil {
		return ns, nil
	}

	if !apierrs.IsNotFound(err) {
		return nil, err
	}

	// probably filling it for Get does not make harm
	ns.ObjectMeta.Name = namespace
	// *ns = corev1.Namespace{
	// 	ObjectMeta: metav1.ObjectMeta{
	// 		Name: namespace,
	// 	},
	// }

	if err = ApplyMetaOptions(ns, metaOptions...); err != nil {
		return nil, err
	}

	err = cli.Create(context.TODO(), ns)
	// it may be different ns already, Get/Create not atomic, but the same in the original code
	return ns, client.IgnoreAlreadyExists(err)
}
``` 

is more clear, but the patch is really nice, no objections to it :)

Reply:
Yeah that might be even nicer, I missed the "merge train". Will get back to it after my PTO :)

Reply:
> Yeah that might be even nicer, I missed the "merge train". Will get back to it after my PTO :)

Probably other functions in the file ask for some attention as well :)